pkg/hive/cell/health.go

```diff
@@ -0,0 +1,249 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cell
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/cilium/cilium/pkg/inctimer"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/stream"
+
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+// How big the queue buffer for health status updates should be.
+// Insufficient will result in reporter status updates blocking while
+// waiting for the queue to be processed.
+const updatesBufferSize = 2048
+
+// Level denotes what kind an update is.
+type Level string
+
+const (
+	StatusUnknown  Level = "Unknown"
+	StatusStopped  Level = "Stopped"
+	StatusDegraded Level = "Degraded"
+	StatusOK       Level = "OK"
+)
+
+// StatusReporter provides a method of declaring a Modules health status.
+type StatusReporter interface {
+	// OK declares that a Module has achieved a desired state and has not entered
+	// any unexpected or incorrect states.
+	// Modules should only declare themselves as 'OK' once they have stabilized,
+	// rather than during their initial state. This should be left to be reported
+	// as the default "unknown" to denote that the module has not reached a "ready"
+	// health state.
+	OK(status string)
+
+	// Stopped reports that a module has completed, and will no longer report any
+	// health status.
+	Stopped(reason string)
+
+	// Degraded declares that a module has entered a degraded state.
+	// This means that it may have failed to provide it's intended services, or
+	// to perform it's desired task.
+	Degraded(reason string, err error)
+}
+
+// Health provides exported functions for accessing health status data.
+// As well, provides unexported functions for use during module apply.
+type Health interface {
+	// All returns a copy of all module statuses.
+	// This includes unknown status for modules that have not reported a status yet.
+	All() []Status
+
+	// Get returns a copy of a modules status, by module ID.
+	// This includes unknown status for modules that have not reported a status yet.
+	Get(string) *Status
+
+	// forModule creates a moduleID scoped reporter handle.
+	forModule(string) StatusReporter
+}
+
+// Update is an event that denotes the change of a modules health state.
+type Update struct {
```

Comment:
Makes the composed Status Update data available.

Reply:
This seems odd. I would favor defaulting to unexporting as much as we can given the api is in flux. 
Also it seems this is only used within the package? otherwise provide a constructor if needed... 
If the tests used an external package we would gain a better appreciation of the apis we're trying to expose.